csharp/src/Apache.Arrow.Adbc/Core/Interop.cs

```diff
@@ -0,0 +1,595 @@
+            {
+                return AdbcStatusCode.UnknownError;
+            }
+
+            GCHandle gch = GCHandle.FromIntPtr(nativeStatement.private_data);
+            AdbcStatement stub = (AdbcStatement)gch.Target;
+            stub.Dispose();
+            gch.Free();
+            nativeStatement.private_data = IntPtr.Zero;
+            return AdbcStatusCode.Success;
+        }
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+    public struct NativeAdbcDatabase
+    {
+        public IntPtr private_data;
+        public IntPtr private_driver;
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+    public struct NativeAdbcConnection
+    {
+        public IntPtr private_data;
+        public IntPtr private_driver;
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+    public struct NativeAdbcStatement
+    {
+        public IntPtr private_data;
+        public IntPtr private_driver;
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+    public struct NativeAdbcPartitions
+    {
+        /// \brief The number of partitions.
+        public IntPtr num_partitions;
+
+        /// \brief The partitions of the result set, where each entry (up to
+        ///   num_partitions entries) is an opaque identifier that can be
+        ///   passed to AdbcConnectionReadPartition.
+        public IntPtr partitions;
+
+        /// \brief The length of each corresponding entry in partitions.
+        public IntPtr partition_lengths;
+
+        /// \brief Opaque implementation-defined state.
+        /// This field is NULLPTR iff the connection is unintialized/freed.
+        public IntPtr private_data;
+
+        /// \brief Release the contained partitions.
+        ///
+        /// Unlike other structures, this is an embedded callback to make it
+        /// easier for the driver manager and driver to cooperate.
+        public IntPtr release; // void (*release)(struct AdbcPartitions* partitions);
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+    public struct NativeAdbcError
+    {
+        /// \brief The error message.
+        public IntPtr message;
+
+        /// \brief A vendor-specific error code, if applicable.
+        public int vendor_code;
+
+        /// \brief A SQLSTATE error code, if provided, as defined by the
+        ///   SQL:2003 standard.  If not set, it should be set to
+        ///   "\0\0\0\0\0".
+        public char sqlstate0;
+        public char sqlstate1;
+        public char sqlstate2;
+        public char sqlstate3;
+        public char sqlstate4;
+
+        /// \brief Release the contained error.
+        ///
+        /// Unlike other structures, this is an embedded callback to make it
+        /// easier for the driver manager and driver to cooperate.
+        public IntPtr release; // void (*release)(struct AdbcError* error);
+    };
+
+
+    [StructLayout(LayoutKind.Sequential)]
+    public struct NativeAdbcDriver
+    {
+        public IntPtr private_data;
+        public IntPtr private_manager;

```

Comment:
Since this is `public` API - let's follow the same pattern that Apache.Arrow C API uses. Let's not use `IntPtr` in public API for things that are pointers. See https://github.com/apache/arrow/pull/34133#discussion_r1124941545 for more information/reasoning.

Reply:
working through these